Add documentation on how to restrict cross namespace associations (#2635)
elastic/cloud-on-k8s: docs/operating-eck/restrict-cross-namespace-associations.asciidoc

----

Then, create a `ServiceAccount` and a `RoleBinding` in the Elasticsearch namespace to allow any resource using the `ServiceAccount` to associate with the Elasticsearch cluster:

[source,sh]
----
> kubectl create serviceaccount associated-resource-sa
----

[source,yaml]
----
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: allow-associated-resource-from-remote-namespace
  namespace: elasticsearch-ns
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: elasticsearch-association
subjects:
  - kind: ServiceAccount
    name: associated-resource-sa
    namespace: associated-resource-ns
----

Finally set the `serviceAccountName` field in the associated resource to specify which `ServiceAccount` is used to create the association:

[source,yaml,subs="attributes"]
----
apiVersion: kibana.k8s.elastic.co/{eck_crd_version}
kind: Kibana
metadata:
  name: associated-resource
  namespace: associated-resource-ns
spec:
 ...
  elasticsearchRef:
    name: "elasticsearch-sample"
    namespace: "elasticsearch-ns"
  # Service account used by this resource to get access to an Elasticsearch cluster
  serviceAccountName: associated-resource-sa
----

> In the above example, `associated-resource` can be of any `Kind` that requires an association to be created (e.g. `Kibana` or the `APMServer`).

Reviewer: avoid latinisms
e.g. => for ex.

You can find https://github.com/elastic/cloud-on-k8s/blob/master/config/recipes/associations-rbac/apm_es_kibana_rbac.yaml[a complete example in the ECK GitHub repository].

> NOTE: If the `serviceAccountName` is not set then ECK uses the `ServiceAccount` called `default`

Reviewer: ```suggestion
NOTE: If the `serviceAccountName` is not set, then ECK uses the `ServiceAccount` called `default`.
```

:page_id: restrict-cross-namespace-associations
ifdef::env-github[]
****
link:https://www.elastic.co/guide/en/cloud-on-k8s/master/restrict-cross-namespace-associations.html[View this document on the Elastic website]
****
endif::[]
[id="{p}-{page_id}"]

> = Restrict cross namespace resource associations

Reviewer 2: ```suggestion
= Restrict cross-namespace resource associations
```

By default, associations can be created between any resources as long as they are deployed in namespaces managed by the same ECK instance.

This is true, for example, when using the `elasticsearchRef` field to automatically configure and establish a trust relationship between Kibana and Elasticsearch. If a user can create a Kibana instance in a namespace managed by ECK, then this Kibana instance can be associated with any Elasticsearch cluster deployed in another namespace managed by the same ECK instance, even if the user only has permissions to edit Kibana and not the Elasticsearch resource.

> 

Reviewer 2: I like Charith's suggestion but would change the last sentence to "Association succeeds even if the user does not have access to one of the namespaces, or only has access to Kibana resources and not Elasticsearch resources." or something similar

This section describes how you can restrict the associations that can be created between resources managed by ECK.

> == Enabling cross namespace association restriction

Reviewer 3: ```suggestion
== Enabling cross-namespace association restrictions
```

This feature is disabled by default. To enable it, start the operator with the `--enforce-rbac-on-refs` flag.

> NOTE: This feature only provides an access control for resources deployed across two different namespaces. Associations can still be freely created between resources deployed in the same namespace.

Reviewer 3: ```suggestion
NOTE: This feature only provides access control for resources deployed across two different namespaces. Associations can still be freely created between resources deployed in the same namespace.
```

Reviewer 3: @charith-elastic 's suggestion is better

Once enabled, associations can be created as long as the `ServiceAccount` specified in the associated resources can execute an HTTP `get` request against the referenced Elasticsearch object.

IMPORTANT: ECK removes any association that is not allowed. If you have existing associations, do not enable this feature without creating the required `Roles` and `RoleBindings` as described in the following sections.

> == Allowing associations to be created

Reviewer 4: Probabaly don't need this heading at all? 
```suggestion

```

First create a `ServiceAccount`:

[source,sh]
----
> kubectl create serviceaccount associated-resource-sa
----

> 

Reviewer 4: I think that it might be better to move the section on creating the cluster role and associating it with the service account here because those are one-off tasks that are closely related.

Then set the `serviceAccountName` field in the associated resource to specify which `ServiceAccount` is used to create the association:

[source,yaml,subs="attributes"]
----

> apiVersion: <kind>.k8s.elastic.co/{eck_crd_version}

Reviewer 2: I wonder if it might be better to use a concrete type in the example, and call out that it also applies to the other

kind: <Kind>
metadata:
  name: associated-resource
  namespace: associated-resource-ns
spec:
 ...
  elasticsearchRef:
    name: "elasticsearch-sample"
    namespace: "elasticsearch-ns"
  # Service account used by this resource to get access to an Elasticsearch cluster
  serviceAccountName: associated-resource-sa
----

In the above example, `associated-resource` can be of any `Kind` that requires an association to be created (e.g. `Kibana` or the `APMServer`).
You can find https://github.com/elastic/cloud-on-k8s/blob/master/config/recipes/associations-rbac/apm_es_kibana_rbac.yaml[a complete example in the ECK GitHub repository].

NOTE: If the `serviceAccountName` is not set then ECK uses the `ServiceAccount` called `default`

> Create a `ClusterRole` to allow an HTTP `get` request to be run against Elasticsearch objects:

Reviewer 3: ```suggestion
Create a `ClusterRole` to allow HTTP `get` requests to be run against Elasticsearch objects:
```

[source,yaml]
----
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: elasticsearch-association
rules:
  - apiGroups:
      - elasticsearch.k8s.elastic.co
    resources:
      - elasticsearches
    verbs:
      - get
----

> Then, to allow the `ServiceAccount` of the associated resource to use this role create the following `RoleBinding`:

Reviewer 4: ```suggestion
Then, create a `RoleBinding` in the Elasticsearch namespace to allow any resource presenting the correct service account token to associate with the Elasticsearch cluster:
```

Reviewer 2: I wonder if "presenting the correct service account token" is overly specific. Maybe just:
>Then, create a `RoleBinding` in the Elasticsearch namespace to allow any resource using the `ServiceAccount` to associate with the Elasticsearch cluster

?

[source,yaml]
----
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: allow-associated-resource-from-remote-namespace
  namespace: elasticsearch-ns
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: elasticsearch-association
subjects:
  - kind: ServiceAccount
    name: associated-resource-sa
    namespace: associated-resource-ns
----

> The associated resource `associated-resource` is now allowed to create an association with any Elasticsearch cluster in the namespace `elasticsearch-ns`.

Reviewer 4: ```suggestion

```

Reviewer 2: I actually think I like having the summary there but don't feel strongly either way
